darkcluster/src/main/java/com/linkedin/darkcluster/util/ExpiringCircularBuffer.java

```diff
@@ -0,0 +1,143 @@
+ * @param <E>
+ */
+@SuppressWarnings("serial")
+public class ExpiringCircularBuffer<E> extends ArrayList<E> implements Queue<E> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ExpiringCircularBuffer.class);
+  private Duration _ttl;
+  private ArrayList<Instant> _ttlBuffer;
+  private volatile int _readerPosition;
+  private volatile int _writerPosition;
+
+  public ExpiringCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit) {
+    super();
+    setCapacity(capacity);
+    setTtl(ttl, ttlUnit);
+  }
+
+  public synchronized boolean add(E toAdd) {
+    this.set(_writerPosition, toAdd);
+    _ttlBuffer.set(_writerPosition, Instant.now());
+    bumpWriterPosition();
+    return true;
+  }
+
+  /**
+   * TODO: javadocs
+   * @return
+   */
+  public synchronized E element() {
+    LOGGER.info("Trying a get. Reader position is {}", _readerPosition);
+    E e = this.get(_readerPosition);
+    Instant ttl = _ttlBuffer.get(_readerPosition);
+    if (e == null) {
+      if (_readerPosition == 0) {
+        throw new NoSuchElementException("Buffer is empty");
+      } else {
+        LOGGER.info("Hit a hole in the buffer, bump readerPosition and retry.");
+        bumpReaderPosition();
+        return element();
+      }
+    } else if (Duration.between(ttl, Instant.now()).compareTo(_ttl) > 0) {
+      if (_readerPosition != 0 || (_readerPosition == 0 && this.size() == 1)) {
+        LOGGER.info("purging object from buffer after expiring ttl");
+        this.set(_readerPosition, null);
+        _ttlBuffer.set(_readerPosition, null);
+      }
+      bumpReaderPosition();
+      return element();
+    }
+    bumpReaderPosition();
+    return e;
+  }
+
+  public boolean offer(E e) {
+    return add(e);
+  }
+
+  public E remove() {
+    throw new NotImplementedException("ExpiringCircularBuffer does not support removal operations");
+  }
+
+  public E poll() {
+    throw new NotImplementedException("ExpiringCircularBuffer does not support removal operations");
+  }
+
+  public E peek() {
+    try {
+      return element();
+    } catch (NoSuchElementException ex) {
+      return null;
+    }
+  }
+
+  public int size() {
+    return (int) this.stream().filter(Objects::nonNull).count();
+  }
+
+  public boolean isEmpty() {
+    return this.size() == 0;
+  }
+
+  private synchronized void bumpReaderPosition() {
+    _readerPosition = (_readerPosition + 1) % super.size();
+    LOGGER.info("reader position is now {}", _readerPosition);
+  }
+
+  private synchronized void bumpWriterPosition() {
+    _writerPosition = (_writerPosition + 1) % super.size();
+    LOGGER.info("writer position is now {}", _writerPosition);
+  }
+
+  public synchronized void setCapacity(int capacity) {
+    if (capacity < 1) {
+      throw new IllegalArgumentException("capacity can't be less than 1.");
+    }
+    super.clear();
+    super.addAll(Collections.nCopies(capacity, null));
+    _ttlBuffer = new ArrayList<>(Collections.nCopies(capacity, null));
+  }
```

Comment:
Since the buffer does not allow dynamic resizing, do we really need a setCapacity() public method? I think it suffices to initialize an ArrayList in the constructor.

Reply:
This is tricky. We need to be able to allow upstream users the ability to supply their own buffer (LI has its own reasons for wanting to do this for example). After it's supplied though, there exists the need to change the size of that buffer through configuration while the application is running. If the upstream users didn't need to supply their own implementation, we could just re-instantiate a rest.li-only component here with updated constructor params.